<a href="https://colab.research.google.com/github/sosarodrigox/grupo13_pp2/blob/main/ProyectoIntegrador_Grupo13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Instituto Superior Politécnico Córdoba

## Tecnicatura Superior en Ciencia de Datos e Inteligencia Artificial

### Cátedra: Práctica Profesionalizante II

**Docente**: Facundo Oliva Cuneo

**Año**: 2024
  
**Trabajo Práctico Integrador**

## Grupo 13
### Integrantes:

- **Ingaramo, Ma. Eugenia**  
  [eugenia.ingaramo@gmail.com](mailto:eugenia.ingaramo@gmail.com)
- **Lonardi, Pablo**  
  [lonardipablo@gmail.com](mailto:lonardipablo@gmail.com)
- **Margheim, Carolina**  
  [caro08.m@gmail.com](mailto:caro08.m@gmail.com)
- **Sosa, Rodrigo**  
  [sosarodrigox@gmail.com](mailto:sosarodrigox@gmail.com)
- **Zenere, Mauricio**  
  [zeneremauricio@gmail.com](mailto:zeneremauricio@gmail.com)

### CRISP-DM 1. Compresión del negocio:

### CRISP-DM 2. Compresión de los datos:

In [87]:
import pandas as pd

# URL del archivo en formato 'Raw' en GitHub
url = "https://raw.githubusercontent.com/sosarodrigox/grupo13_pp2/main/data/data.csv"

# Cargar los datos en un DataFrame
data = pd.read_csv(url)

# Mostrar las primeras filas para confirmar la carga correcta
print(data.head())

          isbn13      isbn10           title subtitle  \
0  9780002005883  0002005883          Gilead      NaN   
1  9780002261982  0002261987    Spider's Web  A Novel   
2  9780006163831  0006163831    The One Tree      NaN   
3  9780006178736  0006178731  Rage of angels      NaN   
4  9780006280897  0006280897  The Four Loves      NaN   

                           authors                     categories  \
0               Marilynne Robinson                        Fiction   
1  Charles Osborne;Agatha Christie  Detective and mystery stories   
2             Stephen R. Donaldson               American fiction   
3                   Sidney Sheldon                        Fiction   
4              Clive Staples Lewis                 Christian life   

                                                                                               thumbnail  \
0  http://books.google.com/books/content?id=KQZCPgAACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api   
1  http://books.google.com

In [88]:
# Configurar opciones de visualización
pd.set_option("display.max_columns", None)  # Todas las columnas
pd.set_option("display.max_rows", 10)  # 0 filas

# DataFrame
data

isbn13      isbn10                      title  \
0     9780002005883  0002005883                     Gilead   
1     9780002261982  0002261987               Spider's Web   
2     9780006163831  0006163831               The One Tree   
3     9780006178736  0006178731             Rage of angels   
4     9780006280897  0006280897             The Four Loves   
...             ...         ...                        ...   
6805  9788185300535  8185300534                  I Am that   
6806  9788185944609  8185944601       Secrets Of The Heart   
6807  9788445074879  8445074873             Fahrenheit 451   
6808  9789027712059  9027712050   The Berlin Phenomenology   
6809  9789042003408  9042003405  'I'm Telling You Stories'   

                                            subtitle  \
0                                                NaN   
1                                            A Novel   
2                                                NaN   
3                                                NaN   
4                                                NaN   
...                                              ...   
6805             Talks with Sri Nisargadatta Maharaj   
6806                                             NaN   
6807                                             NaN   
6808                                             NaN   
6809  Jeanette Winterson and the Politics of Reading   

                                           authors  \
0                               Marilynne Robinson   
1                  Charles Osborne;Agatha Christie   
2                             Stephen R. Donaldson   
3                                   Sidney Sheldon   
4                              Clive Staples Lewis   
...                                            ...   
6805  Sri Nisargadatta Maharaj;Sudhakar S. Dikshit   
6806                                 Khalil Gibran   
6807                                  Ray Bradbury   
6808                 Georg Wilhelm Friedrich Hegel   
6809                        Helena Grice;Tim Woods   

                         categories  \
0                           Fiction   
1     Detective and mystery stories   
2                  American fiction   
3                           Fiction   
4                    Christian life   
...                             ...   
6805                     Philosophy   
6806                      Mysticism   
6807                   Book burning   
6808                        History   
6809             Literary Criticism   

                                                                                                            thumbnail  \
0               http://books.google.com/books/content?id=KQZCPgAACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api   
1               http://books.google.com/books/content?id=gA5GPgAACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api   
2               http://books.google.com/books/content?id=OmQawwEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api   
3               http://books.google.com/books/content?id=FKo2TgANz74C&printsec=frontcover&img=1&zoom=1&source=gbs_api   
4               http://books.google.com/books/content?id=XhQ5XsFcpGIC&printsec=frontcover&img=1&zoom=1&source=gbs_api   
...                                                                                                               ...   
6805            http://books.google.com/books/content?id=Fv_JPQAACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api   
6806            http://books.google.com/books/content?id=XcrVpL1byOkC&printsec=frontcover&img=1&zoom=1&source=gbs_api   
6807                                                                                                              NaN   
6808  http://books.google.com/books/content?id=Vy7SkjvmfLMC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api   
6809  http://books.google.com/books/content?id=2lVyRH0TWugC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api   

                        

### Analisis de datos EDA:

In [89]:
# Obtener informacion del dataset recibido:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6810 entries, 0 to 6809
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   isbn13          6810 non-null   int64  
 1   isbn10          6810 non-null   object 
 2   title           6810 non-null   object 
 3   subtitle        2381 non-null   object 
 4   authors         6738 non-null   object 
 5   categories      6711 non-null   object 
 6   thumbnail       6481 non-null   object 
 7   description     6548 non-null   object 
 8   published_year  6804 non-null   float64
 9   average_rating  6767 non-null   float64
 10  num_pages       6767 non-null   float64
 11  ratings_count   6767 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 638.6+ KB


### Resumen del Dataset de Libros

El dataset analizado consta de **6,810 entradas**, cada una representa un libro único. Estos están descritos mediante **12 atributos**, que incluyen tanto identificadores numéricos como descriptivos. A continuación se detalla la estructura y las características principales del dataset:

### Identificadores de Libro
- `isbn13`: Cada libro tiene un identificador ISBN de 13 dígitos, proporcionado para todos los libros (6810 entradas no nulas).
- `isbn10`: Un identificador ISBN alternativo de 10 dígitos para cada libro, también completamente poblado.

### Títulos y Autores
- `title`: El título del libro, sin entradas nulas en el dataset.
- `subtitle`: Subtítulos proporcionados para 2,381 libros, sugiriendo que no todos los libros tienen un subtítulo explícito.
- `authors`: Nombres de los autores, presentes para 6,738 libros, indicando una pequeña proporción de entradas sin esta información. Esto podría representar un problema a futuro.

### Categorización y Descripción
- `categories`: Categorías o géneros bajo los cuales se clasifican los libros, con 6,711 libros categorizados. Hay casi 100 libros no categorizados lo cual podria generar conflictos a la hora de realizar recomendaciones.
- `description`: Descripciones textuales detalladas disponibles para 6,548 libros, proporcionando un contexto valioso sobre el contenido y el tema de cada libro y que podria utilizarce para las recomendaciones a clientes de la biblioteca.

### Aspectos Visuales y Editoriales
- `thumbnail`: URLs de imágenes de portada para 6,481 libros, lo que ayuda en la visualización y el marketing digital del libro. Se testeó valores al azar y se verificó que algunos links estaban caidos. Evaluar la solución para estos casos.
- `published_year`: Año de publicación registrado para 6,804 libros, ofreciendo una perspectiva sobre la temporalidad de las publicaciones.

### Calificaciones y Popularidad
- `average_rating`: La calificación promedio asignada por los usuarios, disponible para 6,767 libros, lo que refleja la recepción del libro por parte de los lectores.
- `num_pages`: Número de páginas, también disponible para 6,767 libros, indicando la extensión de los textos.
- `ratings_count`: El número de calificaciones que ha recibido cada libro, también notado para 6,767 libros, que puede ser utilizado para medir la popularidad o el engagement de los lectores.

**Uso del Dataset en Relación con los Objetivos del Proyecto**: Este conjunto de datos ofrece una base sólida para varios componentes esenciales del proyecto. *Es ideal para desarrollar un sistema automatizado de recomendación de libros*, tanto basado en contenido como en colaboración, alineándose con el objetivo de optimizar la experiencia del usuario y aumentar su satisfacción. Sin embargo, para respaldar completamente la iniciativa de mejorar la eficiencia en la gestión de recursos y optimizar la utilización de estos, sería beneficioso enriquecer el dataset con datos adicionales sobre las interacciones de los usuarios con la biblioteca, tales como:

 * Registros de préstamos.
 * Frecuencia de uso.
 * Retroalimentación directa de los usuarios.

Esta información adicional permitiría realizar análisis predictivos más precisos sobre tendencias de publicación y demanda de libros, facilitando una mejor alineación con las necesidades e intereses de los usuarios. Además, la integración y automatización de los registros de disponibilidad y estado de préstamo en el dataset podrían reducir significativamente la carga administrativa del personal, apoyando la eficiencia operativa y proporcionando una base de datos más completa sobre la cual tomar decisiones informadas para la gestión de colecciones y la programación de actividades de la biblioteca.

De todas formas, como se mencionó anteriormente el dataset es **suficiente** y presenta datos de **calidad** para una primera iteración en la construcción de un sistema de recomendación.

### CRISP-DM 3. Preparación de los datos:

Dado que uno de los objetivos es implementar un sistema automatizado de recomendación de libros, vemos que en el dataset contamos con los datos necesarios para identificar un libro y recomendarlo a los usuarios de la biblioteca, procedemos a extraer estas variables importantes y procesarlas.

Las variables seleccionadas son:

* `title` (ítulo del libro)
* `authors` (Autor/es del libro)
* `categories` (Categorías del libro)
* `description` (Descripción del libro)
* `published_year` (Año de publicación)

In [90]:
# Obtenemos las variables de interés para el análisis:

variables_seleccionadas = [
    "title",
    "authors",
    "categories",
    "description",
    "published_year",
]

print(data[variables_seleccionadas].info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6810 entries, 0 to 6809
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           6810 non-null   object 
 1   authors         6738 non-null   object 
 2   categories      6711 non-null   object 
 3   description     6548 non-null   object 
 4   published_year  6804 non-null   float64
dtypes: float64(1), object(4)
memory usage: 266.1+ KB
None


Antes de construir el sistema de recomendación, debemos preprocesar los datos. Esto incluye limpieza de texto, manejo de valores faltantes y tokenización.

In [91]:
# Revisar si hay valores nulos en las variables seleccionadas y reemplazarlos por un valor por defecto
for feature in variables_seleccionadas:
    data[feature] = data[feature].fillna("")

# Revisar si hay valores nulos en las variables seleccionadas
print(data[variables_seleccionadas].info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6810 entries, 0 to 6809
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   title           6810 non-null   object
 1   authors         6810 non-null   object
 2   categories      6810 non-null   object
 3   description     6810 non-null   object
 4   published_year  6810 non-null   object
dtypes: object(5)
memory usage: 266.1+ KB
None


Ahora que no tenemos valores nulos, combinamos las columnas de interés en una nueva columna llamada `libros`. El sistema de recomendación utilizará esta columna combinada para encontrar similitudes entre los libros de la biblioteca.

In [92]:
# Combinar las columnas de interés en una sola columna
libros = (
    data["title"]
    + " "
    + data["categories"]
    + " "
    + data["authors"]
    + " "
    + f"{data['published_year']}"
)
pd.set_option("display.max_colwidth", None)
libros.head(5)

0                                             Gilead Fiction Marilynne Robinson 0       2004.0\n1       2000.0\n2       1982.0\n3       1993.0\n4       2002.0\n         ...  \n6805    1999.0\n6806    1993.0\n6807    2004.0\n6808    1981.0\n6809    1998.0\nName: published_year, Length: 6810, dtype: object
1    Spider's Web Detective and mystery stories Charles Osborne;Agatha Christie 0       2004.0\n1       2000.0\n2       1982.0\n3       1993.0\n4       2002.0\n         ...  \n6805    1999.0\n6806    1993.0\n6807    2004.0\n6808    1981.0\n6809    1998.0\nName: published_year, Length: 6810, dtype: object
2                            The One Tree American fiction Stephen R. Donaldson 0       2004.0\n1       2000.0\n2       1982.0\n3       1993.0\n4       2002.0\n         ...  \n6805    1999.0\n6806    1993.0\n6807    2004.0\n6808    1981.0\n6809    1998.0\nName: published_year, Length: 6810, dtype: object
3                                         Rage of angels Fiction Sidney Sheldon 0 

### CRISP-DM 4. Modelado:

Para crear el sistema de recomendación basado en contenido, utilizaremos la técnica de **TF-IDF** (Term Frequency-Inverse Document Frequency). Esta técnica se utiliza para convertir texto en vectores numéricos, que luego se pueden utilizar para calcular la similitud entre los textos. TF-IDF otorga más peso a los términos que son importantes en un documento específico y menos peso a los términos comunes.

* Frecuencia de Término (TF): Mide cuántas veces aparece un término en un documento.
* Frecuencia Inversa de Documento (IDF): Mide la importancia de un término en el conjunto de documentos. Los términos comunes que aparecen en muchos documentos reciben un peso menor.

In [93]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Crear un vectorizador TF-IDF
vector_tfidf = TfidfVectorizer()

vector_caracteristicas = vector_tfidf.fit_transform(libros)

print(vector_caracteristicas.shape)
print(vector_caracteristicas)

(6810, 10548)
  (0, 6894)	0.06443750710614472
  (0, 2840)	0.06443750710614472
  (0, 143)	0.06443750710614472
  (0, 5557)	0.06443750710614472
  (0, 7629)	0.06443750710614472
  (0, 6655)	0.06443750710614472
  (0, 102)	0.06443750710614472
  (0, 142)	0.06443750710614472
  (0, 91)	0.06443750710614472
  (0, 141)	0.06443750710614472
  (0, 140)	0.06443750710614472
  (0, 139)	0.06443750710614472
  (0, 103)	0.06443750710614472
  (0, 138)	0.06443750710614472
  (0, 110)	0.06443750710614472
  (0, 98)	0.12887501421228945
  (0, 92)	0.06443750710614472
  (0, 108)	0.06443750710614472
  (0, 112)	0.12887501421228945
  (0, 8035)	0.43700046813706117
  (0, 6013)	0.5885172279797457
  (0, 3422)	0.11189605685054443
  (0, 3882)	0.5885172279797457
  (1, 1822)	0.3013956698681646
  (1, 267)	0.30498953565603815
  :	:
  (6809, 10330)	0.33175135989257093
  (6809, 2240)	0.2206043839379868
  (6809, 9492)	0.30218650531542857
  (6809, 10435)	0.26439634843355025
  (6809, 5684)	0.20438577371569946
  (6809, 9065)	0.21319651

**Similitud del Coseno**(Cosine Similarity): La similaridad de coseno es una métrica que se utiliza para medir cuán similares son dos vectores. Se calcula como el coseno del ángulo entre dos vectores en un espacio multidimensional. En términos de recomendación de contenido, esto nos ayuda a medir cuán similares son dos libros en función de sus características textuales vectorizadas.Se calcula tomando el coseno del ángulo entre dos vectores no nulos y su valor oscila entre -1 y 1, donde:

* Un valor de 1 indica que los vectores son idénticos.
* Un valor de 0 indica que los vectores son ortogonales (no tienen similitudes).
* Un valor de -1 indica que los vectores son diametralmente opuestos.
* La diagonal principal de la matriz siempre estará compuesta por unos (1,1,1,...,1), ya que cada libro es idéntico a sí mismo.

En la función cosine_similarity() pasamos el `vector_caracteristicas` dos veces indicando que estamos calculando la similaridad de cada libro con todos los demás libros (incluido él mismo) para generar una matriz de similaridad. En esta matriz, la entrada [i, j] representa la similaridad entre el libro i y el libro j

Ejemplo Visual: Si tenemos tres libros (A, B, C), la matriz de similaridad resultante tendría este aspecto:

        A    B    C
    A  1.0  0.8  0.3
    B  0.8  1.0  0.5
    C  0.3  0.5  1.0

Aquí, cada celda [i, j] representa la similaridad entre el libro i y el libro j, donde el valor de 1.0 en la diagonal indica que un libro es completamente similar a sí mismo.


In [94]:
from sklearn.metrics.pairwise import cosine_similarity

# Obtener la similitud coseno entre los vectores de características:
similitud = cosine_similarity(vector_caracteristicas, vector_caracteristicas)

print(similitud)

[[1.         0.08005311 0.11641123 ... 0.08492469 0.08320896 0.07576596]
 [0.08005311 1.         0.08011208 ... 0.06549297 0.06416982 0.10645452]
 [0.11641123 0.08011208 1.         ... 0.08498725 0.09570742 0.07582177]
 ...
 [0.08492469 0.06549297 0.08498725 ... 1.         0.06807484 0.06198557]
 [0.08320896 0.06416982 0.09570742 ... 0.06807484 1.         0.06073329]
 [0.07576596 0.10645452 0.07582177 ... 0.06198557 0.06073329 1.        ]]


### CRISP-DM 5. Evaluación:

En esta fase, evaluaremos la efectividad del sistema de recomendación. Este paso implica probar el sistema con ejemplos concretos y verificar si las recomendaciones son relevantes.

Pasos a seguir:

* Crear una lista con todos los títulos de libros en el dataset.
* Obtener el nombre del libro del usuario.
* Encontrar la coincidencia más cercana para el nombre del libro dado por el usuario.
* Encontrar el índice del libro con el título coincidente.
* Obtener una lista de libros similares basados en el puntaje de similitud.
* Ordenar los libros basados en su puntaje de similitud.
* Imprimir los nombres de libros similares basados en el índice.

In [95]:
# Crear una lista con todos los títulos de libros en el dataset
titulos = data["title"].tolist()
print(titulos)

['Gilead', "Spider's Web", 'The One Tree', 'Rage of angels', 'The Four Loves', 'The Problem of Pain', 'An Autobiography', 'Empires of the Monsoon', 'The Gap Into Madness', 'Master of the Game', 'If Tomorrow Comes', "Assassin's Apprentice", 'Warhost of Vastmark', 'The Once and Future King', 'Murder in LaMut', 'Jimmy the Hand', 'Well of Darkness', 'Witness for the Prosecution & Selected Plays', 'The Little House', 'Mystical Paths', 'Glittering Images', 'Glamorous Powers', 'The Mad Ship', 'Post Captain', 'The Reverse of the Medal', 'Miss Marple', 'The Years of Rice and Salt', 'Spares', 'Gravity', 'The Wise Woman', "Girls' Night in", 'The White Album', "The Bonesetter's Daughter", 'The Lexus and the Olive Tree', "'Tis", 'Ocean Star Express', 'A Small Pinch of Weather', 'The Princess of the Chalet School', 'Koko', 'Tree and Leaf', 'Partners in Crime', 'Murder in Mesopotamia', 'The Lord of the Rings, the Return of the King', 'All Families are Psychotic', 'Death in the Clouds', 'Appointment w

In [96]:
# Obtener el nombre del libro del usuario.
def obtener_libro_usuario():
    libro_usuario = input("Ingrese el título de un libro: ")
    return libro_usuario

Para encontrar las coincidencias usamos la librerías difflib, esta proporciona clases y funciones para comparar secuencias, y es particularmente útil para tareas de procesamiento de texto y análisis de diferencias. En este caso, usamos la función get_close_matches() para encontrar las coincidencias más cercanas al título del libro dado por el usuario.

In [97]:
from difflib import get_close_matches

# Primero obtenemos el nombre del libro del usuario:
libro_usuario = obtener_libro_usuario()


# Encontrar la coincidencia más cercana para el nombre del libro dado por el usuario.
def encontrar_coincidencia(titulo_ingresado):
    return get_close_matches(titulo_ingresado, titulos)


coincidencia = encontrar_coincidencia(libro_usuario)

print(coincidencia)

['Dracula', 'Dracula', 'Dracula']


In [98]:
# Encontrar el índice del libro con el título coincidente.
coincidencia_index = coincidencia[0]
indice_libro = data[data["title"] == coincidencia_index].index[0]

In [99]:
# Obtener una lista de libros similares basados en el puntaje de similitud.
puntaje_similitud = list(enumerate(similitud[indice_libro]))
print(puntaje_similitud)

[(0, 0.06940174955517588), (1, 0.05352185441007998), (2, 0.06945287619558656), (3, 0.06451217578843185), (4, 0.057282434778645974), (5, 0.05327991035875435), (6, 0.0711744770317461), (7, 0.05356250482970559), (8, 0.18914509718253625), (9, 0.06037323008090995), (10, 0.054293973540840076), (11, 0.06133529279556172), (12, 0.0600578538857701), (13, 0.04750367949806789), (14, 0.04993946491250528), (15, 0.05850101422076297), (16, 0.06029642849632127), (17, 0.0515103277904796), (18, 0.06540696090109199), (19, 0.059998786162562136), (20, 0.059544215912171565), (21, 0.05716467417159121), (22, 0.060344681324006465), (23, 0.20972774927581997), (24, 0.05798835286921798), (25, 0.05582272140975911), (26, 0.05256297870657384), (27, 0.05475001966357443), (28, 0.07112178039603335), (29, 0.05915755744741476), (30, 0.048939074376327996), (31, 0.05787594566923228), (32, 0.058780915536352236), (33, 0.051962088713156204), (34, 0.06346227303767728), (35, 0.05179961575818634), (36, 0.05111412642487227), (37, 

In [100]:
# Ordenar los libros basados en su puntaje de similitud.
libros_similares_ordenados = sorted(puntaje_similitud, key=lambda x: x[1], reverse=True)
print(libros_similares_ordenados)

[(859, 1.0000000000000002), (2447, 0.7961758733666621), (4443, 0.7961758733666621), (4933, 0.6077556185533894), (4946, 0.4584548910996259), (5865, 0.39170626302840605), (2501, 0.28880682930055107), (5892, 0.28464238074444725), (2836, 0.2678835540968578), (3184, 0.24811569160598562), (4290, 0.2464113066381569), (80, 0.23462008901905634), (58, 0.23460600525362277), (871, 0.23066638219320124), (4291, 0.22349060739514526), (2555, 0.22003996432800887), (6591, 0.2162359901201435), (1741, 0.21016201128821238), (23, 0.20972774927581997), (57, 0.20967592778793415), (5622, 0.20899282583244494), (6565, 0.20582806717452762), (40, 0.20477673894503065), (1735, 0.20400582253952143), (4573, 0.19993388192419945), (4570, 0.19778976400579576), (4572, 0.19407112102136007), (3711, 0.19296363216805293), (46, 0.19022096671076644), (4539, 0.18957128935572448), (8, 0.18914509718253625), (65, 0.1869393918289997), (4456, 0.18646434280854324), (4596, 0.1863603508055093), (1335, 0.1843630999959195), (875, 0.183725

### CRISP-DM 6. Despliegue: